workbook/chapter09.md

```diff
@@ -0,0 +1,213 @@
+
+### Create a new module 
+
+To start this you will need to create a new module, to do this follow the instructions on [Chapter 8](chapter8.md). For our setup the module name will be considered *mypid*.
+
+### Create a substitution file
+
+Now we will create in our module a substitution file that uses the `pid_control.db` file as a template file (find the file [here](https://github.com/epics-modules/std/blob/master/stdApp/Db/pid_control.db)). You should create a `pid.substitutions` file within this content:
+
+```
+file "pid_control.db"
+{
+pattern { P,        PID,    INP,        OUT,        LOPR,   HOPR,   DRVL,   DRVH,   PREC,   KP,     KI, KD, SCAN        }
+        { mypid:,   PID1,   pidDemoInp, pidDemoOut, 0,      100,    0,      5,      3,      0.2,    3., 0., ".1 second" }
+}
+```
+
+This file is just an example, and uses as `INP` and `OUT` in existent PVs, but it suffices for our test. Note that there is no hard-code path or variable within the substitution file. 
+
+If you change the `mypid.Makefile` and try to compile this module you should receive a message like this:
+
+```console
+msi: Can't open file 'pid_control.db'
+```
+
+This is because `MSI` has no idea where `pid_control.db` file is. You need to tell the building system where it is. 
+
+### Add std as a dependency
+
+To solve the above, the first step is to set *std* as a dependency. As we've see on previous lesson you should edit `mypid.Makefile` and `CONFIGURE_MODULE`.
+
+On `mypid.Makefile` you should add:
+
+```bash
+REQUIRED += std
+
+ifneq ($(strip $(STD_DEP_VERSION)),)
+std_VERSION=$(STD_DEP_VERSION)
+endif
+```
+
+And add the following line into `CONFIG_MODULE`:
+
+```python
+STD_DEP_VERSION:=3.5.0
+```
+
+### Add std on `USR_DBFLAGS`
+
+To allow that your substitutions file uses db files from *std* you should include the std db folder on `USR_DBFLAGS`. So in `mypid.Makefile` you add this line:
+
+```
+USR_DBFLAGS += -I $(E3_SITEMODS_PATH)/std/$(std_VERSION)/db
+```
+
+This line will tell to `MSI` where find the `pid_control.db`.
+
+### Checking if everything is ok
+
+After these changes you can compile your module and you shouldn't see any error.
+
+If you would like to, you can go to your module folder and see the `pid.db` generated file, the file should be at `$(E3_REQUIRE_LOCATION)/siteMods/mypid/master/db/pid.db`
+
+## Example files
+
+The example files showed in this tutorial could be found at 
+[e3-moduleexample](https://gitlab.esss.lu.se/epics-examples/e3-moduleexample.git) and [moduleexample](https://gitlab.esss.lu.se/epics-examples/moduleexample.git). Note that the module name is moduleexample, but the db and substitutions
+files are the same as the ones used in this tutorial.
+
+---
+
+## Assignments
+
+* What does MSI stand for?
+<!-- todo: figure out proper assignments -->
```

Comment:
Here are some ideas:

* The substitution file above only has a very simple substitution pattern: each macro is replaced with a single value. Is there another way to achieve the same result?
* Maybe add a question about what happens if you do e.g.
```bash
require calc
require linconv
```
versus
```bash
require linconv
require calc
```
What happens if you have more than one version of calc installed?
* Perhaps a question about using a .iocsh snippet from another module? (i.e. using `$(std_DIR)` in the startup script)